include/revng/PipelineC/Tracing/Trace.h

```diff
@@ -0,0 +1,229 @@
+#pragma once
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <set>
+#include <string>
+#include <vector>
+
+#include "llvm/Support/Error.h"
+#include "llvm/Support/YAMLTraits.h"
+
+#include "revng/Support/Assert.h"
+#include "revng/Support/Debug.h"
+
+namespace detail {
+
+template<typename T>
+inline T toInt(const llvm::StringRef StrInt) {
+  T Result;
+  revng_assert(!StrInt.getAsInteger(10, Result));
+  return Result;
+}
+
+} // namespace detail
+
+namespace revng::tracing {
+
+enum ArgumentState { Undefined, Scalar, Sequence };
+
+struct Argument {
+private:
+  ArgumentState State = Undefined;
+
+public:
+  std::string Scalar;
+  std::vector<std::string> Sequence;
+
+public:
+  void setState(const ArgumentState NewState) {
+    revng_assert(State == Undefined || State == NewState);
+    State = NewState;
+  }
+
+  bool isValid() const { return State != ArgumentState::Undefined; }
+  bool isScalar() const { return State == ArgumentState::Scalar; }
+  bool isSequence() const { return State == ArgumentState::Sequence; }
+
+public:
+  template<typename T>
+  T asInt() const {
+    revng_assert(isScalar());
+    return ::detail::toInt<T>(Scalar);
+  }
+
+  template<typename T>
+  std::vector<T> toIntList() const {
+    revng_assert(isSequence());
+    std::vector<T> Ret;
+    for (auto &Elem : Sequence) {
+      Ret.push_back(::detail::toInt<T>(Elem));
+    }
+    return Ret;
+  }
+
+  std::vector<const char *> toStringList() const {
```

Comment:
1. Everywhere you have `const char *` to mean a string, use `StringRef`.
2. You're returning a vector to `const char *`. Are you sure the lifetime of the pointed data is appropriate?

Reply:
1. Changed signature to `std::vector<llvm::StringRef>` (`utils::toStringList` in `Runner.cpp` still returns the `const char *` version though.
2. The basic "Issue" is that, if a copy of the argument is returned then in `Runner.cpp` there would need to be another rule in `ArgStorage` to store the copied vector